examples/c/tc.bpf.c

```diff
@@ -0,0 +1,34 @@
+// SPDX-License-Identifier: (LGPL-2.1 OR BSD-2-Clause)
+/* Copyright (c) 2022 Hengqi Chen */
+#include <vmlinux.h>
+#include <bpf/bpf_endian.h>
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_tracing.h>
+
+#define TC_ACT_OK	0
+#define ETH_P_IP	0x0800		/* Internet Protocol packet	*/
+
+SEC("tc")
+int tc_ingress(struct __sk_buff *ctx)
+{
+	void *data_end = (void *)(__u64)ctx->data_end;
+	void *data = (void *)(__u64)ctx->data;
+	struct ethhdr *l2;
+	struct iphdr *l3;
+
+	if (ctx->protocol != bpf_htons(ETH_P_IP))
+		return TC_ACT_OK;
```

Comment:
just curious why return TC_ACT_OK to pass it to upper stack when it is not a valid IP packet? why not TC_ACT_SHOT ? I know this is just an example to show how to use TC, not how to deal with the packet so I guess it may not be  important? :)

Reply:
Just to avoid interrupting the system running this example. There exists other protocols like ARP.

Reply:
ok make sense here, what about the following two `TC_ACT_OK`?

Reply:
It's used to make the verifier happy.

Reply:
you mean this `if ((void *)(l2 + 1) > data_end)` is to make verifier happy, right? but I mean if `if` is true, the data then is  invalid, should we return `TC_ACT_SHOT`? sorry just trying to understand it better from you :)

Reply:
> you mean this `if ((void *)(l2 + 1) > data_end)` is to make verifier happy, right? but I mean if `if` is true, the data then is invalid, should we return `TC_ACT_SHOT`? sorry just trying to understand it better from you :)

If we return `TC_ACT_SHOT`, it means the current packet will be dropped. Meanwhile, the application layer receives no packet and will cause retransmission. 

Thus, the proper way to deal with the invalid packet is to return `TC_ACT_OK`, which means pass the invalid packet to the upper network stack and let the upper layer to make decision on it.

Reply:
got it, this lead me to the https://en.wikipedia.org/wiki/Robustness_principle "be conservative in what you do, be liberal in what you accept from others"